File: essentia-bench.py
#! /usr/bin/env python2.7

import sys
import essentia
import essentia.standard
from time import time
import numpy as np

def benchmark(filename, block, repeats):
    block(filename)

    times = np.zeros(repeats)
    for i in range(repeats):
        print(i)
        t0 = time()
        block(filename)
        t1 = time()
        times[i] = t1 - t0

    print("Essentia-Python\t%s\t%s\t%f\t%f\t%d" % (filename, block.__name__, np.mean(times), np.std(times, ddof = 1), repeats))

_filename = None
_data = None

def data(filename):
    global _data, _filename
    if _data is None or _filename != filename:
        _data = loadwav(filename)
        _filename = filename
    return _data

def nframes(length, framesize, hopsize):
    return (length - framesize) // hopsize + 1

framesize = 1024
hopsize = 256
nbins = (framesize >> 1) + 1

def loadwav(filename):
    return essentia.standard.MonoLoader(filename = filename)()

def loadmp3(filename):
    return essentia.standard.MonoLoader(filename = filename.replace(".wav", ".mp3"))()

def zcr(filename):
    audio = data(filename)
    fc = essentia.standard.FrameCutter(frameSize = framesize, hopSize = hopsize)
    zcr = essentia.standard.ZeroCrossingRate()
    rates = np.zeros((nframes(len(audio), framesize, hopsize) + 10,), dtype=np.float32)
    i = 0

    while True:
        frame = fc(audio)
        if len(frame) == 0:
            break
        rates[i] = zcr(frame)
        i += 1

    return rates


def resample(filename):
    audio = data(filename)
    resample = essentia.standard.Resample(inputSampleRate = 44100, outputSampleRate = 48000)
    return resample(audio)

def stft(filename):
    audio = data(filename)
    fc = essentia.standard.FrameCutter(frameSize = framesize, hopSize = hopsize)
    w = essentia.standard.Windowing(size = framesize, type = "hann")
    fft = essentia.standard.FFT(size = framesize)
    spectrogram = np.zeros((nbins, nframes(len(audio), framesize, hopsize) + 10), dtype=np.complex64)
    i = 0

    while True:
        frame = fc(audio)
        if len(frame) == 0:
            break
        windowed = w(frame)
        spectrum = fft(windowed)
        spectrogram[:, i] = spectrum

    return spectrogram


def mfcc(filename):
    audio = data(filename)
    fc = essentia.standard.FrameCutter(frameSize = framesize, hopSize = hopsize)
    w = essentia.standard.Windowing(size = framesize, type = "hann")
    spec = essentia.standard.Spectrum(size = framesize)
    mfcc = essentia.standard.MFCC(highFrequencyBound = 22050, inputSize = nbins, numberBands = 128, numberCoefficients = 20)

    mfcc_matrix = np.zeros((20, nframes(len(audio), framesize, hopsize) + 1), dtype=np.complex64)
    i = 0

    while True:
        frame = fc(audio)
        if len(frame) == 0:
            break
        windowed = w(frame)
        spectrum = spec(windowed)
        bands, mfcc_vector = mfcc(spectrum)
        mfcc_matrix[:, i] = mfcc_vector

    return mfcc_matrix

if __name__ == "__main__":
    try:
        filename = sys.argv[1]
        task = eval(sys.argv[2])
        assert callable(task)

        if len(sys.argv) >= 4:
            repeats = int(sys.argv[3])
        else:
            repeats = 100

    except:
        print("Usage: %s [filename] [benchmark name] [number of repeats; 100 by default]" % sys.argv[0])

    benchmark(filename, task, repeats)

File: essentia-bench.cpp
#include <essentia/essentia.h>
#include <essentia/algorithmfactory.h>
#include <essentia/essentiamath.h>
#include <sys/time.h>
#include <math.h>

using namespace std;
using namespace essentia;
using namespace essentia::standard;

typedef void (*TASK)(const string&);

#define N_FRAMES(length, framesize, hopsize) (((length) - (framesize) / (hopsize)) + 1)

static double now() {
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return tv.tv_sec * 1.0 + tv.tv_usec / 1e6;
}

static void benchmark(const string& filename, TASK task, int repeats) {
    task(filename);
    double sumdt = 0.0;
    double sumdtdt = 0.0;
    for (int i = 0; i < repeats; i++) {
        cout << i << endl;
        double t0 = now();
        task(filename);
        double t1 = now();
        double dt = t1 - t0;
        sumdt += dt;
        sumdtdt += dt * dt;
    }

    double mean = sumdt / repeats;
    double stdev = sqrt( (sumdtdt / repeats - mean * mean) * repeats / (repeats - 1) );
    cout << "Essentia-C++\t" << filename << "\t" << task << "\t" << mean << "\t" << stdev << "\t" << repeats << endl;
}

static string _filename = "";

int framesize = 1024;
int hopsize = 256;
int nbins = (framesize >> 1) + 1;

static vector<Real> buffer;

vector<Real> loadaudio(const string &filename) {
    vector<Real> buffer;

    Algorithm *audio = AlgorithmFactory::instance().create("MonoLoader", "filename", filename);
    audio->output("audio").set(buffer);
    audio->compute();
    delete audio;

    return buffer;
}

vector<Real>& audio(const string &filename) {
    if (filename != _filename) {
        buffer = loadaudio(filename);
    }
    return buffer;
}

void loadwav(const string &filename) {
    loadaudio(filename);
}

void loadmp3(const string &filename) {
    string mp3name = filename;
    int index = mp3name.rfind(".wav");
    mp3name.replace(index, 4, ".mp3");
    loadaudio(mp3name);
}

void zcr(const string &filename) {
    vector<Real>& buffer = audio(filename);
    vector<Real> frame;
    Real rate;

    AlgorithmFactory& factory = AlgorithmFactory::instance();
    Algorithm *fc = factory.create("FrameCutter", "frameSize", framesize, "hopSize", hopsize);
    Algorithm *zcr = factory.create("ZeroCrossingRate");

    fc->input("signal").set(buffer);
    fc->output("frame").set(frame);
    zcr->input("signal").set(frame);
    zcr->output("zeroCrossingRate").set(rate);

    vector<Real> rates(N_FRAMES(buffer.size(), framesize, hopsize) + 10);

    for (int i = 0; i < rates.size(); i++) {
        fc->compute();
        if (!frame.size()) break;
        zcr->compute();
        rates[i] = rate;
    }

    delete fc, zcr;
}

void resample(const string &filename) {
    vector<Real>& buffer = audio(filename);
    vector<Real> resampled;

    Algorithm *resample = AlgorithmFactory::instance().create("Resample", "inputSampleRate", 44100, "outputSampleRate", 48000);
    resample->input("signal").set(buffer);
    resample->output("signal").set(resampled);
    resample->compute();

    delete resample;
}

void stft(const string &filename) {
    vector<Real>& buffer = audio(filename);
    vector<Real> frame;
    vector<Real> windowed;
    vector<complex<Real>> spectrum;

    AlgorithmFactory& factory = AlgorithmFactory::instance();
    Algorithm *fc = factory.create("FrameCutter", "frameSize", framesize, "hopSize", hopsize);
    Algorithm *w = factory.create("Windowing", "size", framesize, "type", "hann");
    Algorithm *fft = factory.create("FFT", "size", framesize);

    fc->input("signal").set(buffer);
    fc->output("frame").set(frame);
    w->input("frame").set(frame);
    w->output("frame").set(windowed);
    fft->input("frame").set(windowed);
    fft->output("fft").set(spectrum);

    vector<vector<complex<Real>>> spectrogram(N_FRAMES(buffer.size(), framesize, hopsize) + 10);

    for (int i = 0; i < spectrogram.size(); i++) {
        fc->compute();
        if (!frame.size()) break;
        w->compute();
        fft->compute();
        spectrogram[i] = spectrum;
    }

    delete fc, w, fft;
}

void mfcc(const string &filename) {
    vector<Real>& buffer = audio(filename);
    vector<Real> frame;
    vector<Real> windowed;
    vector<Real> spectrum;
    vector<Real> bands;
    vector<Real> mfcc_vector;

    AlgorithmFactory& factory = AlgorithmFactory::instance();
    Algorithm *fc = factory.create("FrameCutter", "frameSize", framesize, "hopSize", hopsize);
    Algorithm *w = factory.create("Windowing", "size", framesize, "type", "hann");
    Algorithm *spec = factory.create("Spectrum", "size", framesize);
    Algorithm *mfcc = factory.create("MFCC", "highFrequencyBound", 22050, "inputSize", nbins, "numberBands", 128, "numberCoefficients", 20);

    fc->input("signal").set(buffer);
    fc->output("frame").set(frame);
    w->input("frame").set(frame);
    w->output("frame").set(windowed);
    spec->input("frame").set(windowed);
    spec->output("spectrum").set(spectrum);
    mfcc->input("spectrum").set(spectrum);
    mfcc->output("mfcc").set(mfcc_vector);
    mfcc->output("bands").set(bands);

    vector<vector<Real>> mfcc_matrix(N_FRAMES(buffer.size(), framesize, hopsize) + 10);

    for (int i = 0; i < mfcc_matrix.size(); i++) {
        fc->compute();
        if (!frame.size()) break;
        w->compute();
        spec->compute();
        mfcc->compute();
        mfcc_matrix[i] = mfcc_vector;
    }

    delete fc, w, spec, mfcc;
}

int main(int argc, char* argv[]) {
    int repeats = 100;
    if (argc < 3) {
        cerr << "Usage " << argv[0] << " [filename] [benchmark name] [number of repeats; 100 by default]" << endl;
        return -1;
    }
    if (argc == 4) {
        repeats = atoi(argv[3]);
    }

    const char* filename = argv[1];
    const char* task = argv[2];

    map<string, TASK> tasks;
    tasks["loadwav"] = loadwav;
    tasks["loadmp3"] = loadmp3;
    tasks["zcr"] = zcr;
    tasks["resample"] = resample;
    tasks["stft"] = stft;
    tasks["mfcc"] = mfcc;

    init();

    if (tasks.count(task) != 0) {
        benchmark(filename, tasks[task], repeats);
    } else {
        cerr << "Task" << task << " not found" << endl;
    }

    return 0;
}
